syntax = "proto2";

package xlb.rpc;

option cc_generic_services = true;

message Error {
    required int32 code = 1;
    required string errmsg = 2;
}

message EmptyRequest {
}

message GeneralResponse {
    required Error error = 1;
}

message Service {
    required string addr = 1;
    required uint32 port = 2;
}

message VirtualServiceRequest {
    required Service svc = 1;
}

message RealServiceRequest {
    required Service virt = 1;
    required Service real = 2;
}

message ServicesResponse {
    required Error error = 1;
    repeated Service list = 2;
}

service Control {
    rpc AddVirtualService (VirtualServiceRequest) returns (GeneralResponse);
    rpc DelVirtualService (VirtualServiceRequest) returns (GeneralResponse);
    rpc ListVirtualService (EmptyRequest) returns (ServicesResponse);

    rpc AttachRealService (RealServiceRequest) returns (GeneralResponse);
    rpc DetachRealService (RealServiceRequest) returns (GeneralResponse);
    rpc ListRealService (VirtualServiceRequest) returns (ServicesResponse);
}
